Copyright 2021 Andrew M. Olney and made available under [CC BY-SA](https://creativecommons.org/licenses/by-sa/4.0) for text and [Apache-2.0](http://www.apache.org/licenses/LICENSE-2.0) for code.

# Preprocessing text: Problem solving

In this session, we'll work with Twitter data from NLTK.
Execute the code below to download `twitter_samples` from `corpora`.
Close the window when the download completes.

In [2]:
import sys
import nltk
sys.argv=['']
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

## Load the data

Start by importing `twitter_samples` from `nltk.corpus` and `nltk`

In [1]:
from nltk.corpus import twitter_samples
import nltk as nltk

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="B@`]CGD1A2?+*muAJugJ">twitter_samples</variable><variable id="]HJ?tF]9lV9p9(|h{fmK">nltk</variable></variables><block type="importFrom" id="Jzjqgf~f}NetD1+v]eM^" x="16" y="64"><field name="libraryName">nltk.corpus</field><field name="libraryAlias" id="B@`]CGD1A2?+*muAJugJ">twitter_samples</field><next><block type="importAs" id="Ub-RL@h_kQXn-V1vNr.e"><field name="libraryName">nltk</field><field name="libraryAlias" id="]HJ?tF]9lV9p9(|h{fmK">nltk</field></block></next></block></xml>

Now get the raw text of `tweets.20150430-223406.json` and store it in variable `tweetsJson`.

Show the substring of `tweetsJson` from its start to `3000`.

In [2]:
tweetsJson = twitter_samples.raw('tweets.20150430-223406.json')

tweetsJson[ : 3000]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="Ws0}bS$/?#R5]:h%]1Q_">tweetsJson</variable><variable id="B@`]CGD1A2?+*muAJugJ">twitter_samples</variable></variables><block type="variables_set" id="Q=bB?13U{5~]-mP$F$RJ" x="41" y="200"><field name="VAR" id="Ws0}bS$/?#R5]:h%]1Q_">tweetsJson</field><value name="VALUE"><block type="varDoMethod" id="5L0KrAA7.ekjugpH0_+("><field name="VAR" id="B@`]CGD1A2?+*muAJugJ">twitter_samples</field><field name="MEMBER">raw</field><data>twitter_samples:raw</data><value name="INPUT"><block type="text" id="@8%*,}Voj8$m/8XlDHQz"><field name="TEXT">tweets.20150430-223406.json</field></block></value></block></value></block><block type="text_getSubstring" id="EiYC?M!TV+R:gf_1a|%9" x="8" y="422"><mutation at1="false" at2="true"></mutation><field name="WHERE1">FIRST</field><field name="WHERE2">FROM_START</field><value name="STRING"><block type="variables_get" id="+JP*z!I697F|rOb:sQ|4"><field name="VAR" id="Ws0}bS$/?#R5]:h%]1Q_">tweetsJson</field></block></value><value name="AT2"><block type="math_number" id="s)?j`{oE+jEk3MU2h-]]"><field name="NUM">300</field></block></value></block></xml>

'{"in_reply_to_status_id": null, "contributors": null, "filter_level": "low", "coordinates": null, "id_str": "593891099434983425", "favorited": false, "created_at": "Thu Apr 30 21:34:06 +0000 2015", "favorite_count": 0, "in_reply_to_status_id_str": null, "retweeted_status": {"in_reply_to_status_id": null, "contributors": null, "filter_level": "low", "coordinates": null, "id_str": "593885295323521025", "favorited": false, "created_at": "Thu Apr 30 21:11:02 +0000 2015", "favorite_count": 3, "in_reply_to_status_id_str": null, "in_reply_to_user_id": null, "in_reply_to_screen_name": null, "place": null, "id": 593885295323521025, "retweeted": false, "truncated": false, "text": "Indirect cost of the UK being in the EU is estimated to be costing Britain \\u00a3170 billion per year! #BetterOffOut #UKIP", "in_reply_to_user_id_str": null, "lang": "en", "retweet_count": 11, "user": {"statuses_count": 4589, "id_str": "929903647", "followers_count": 2973, "listed_count": 39, "geo_enabled": false, "c

This tweet is in a file format called [JSON](https://en.wikipedia.org/wiki/JSON), which stands for JavaScript Object Notation.
JSON is a textual representation of an **object**, and objects can have complex nested structure.
A lot of it is metadata that we don't care about (we only care about the `text` field), so to simplify things, let's use the `twitter_samples.strings` instead of `raw`.

Create a variable `tweets` and set it to `twitter_samples.strings` using `tweets.20150430-223406.json`

Then show the first `10` elements of `tweets`.

In [3]:
tweets = twitter_samples.strings('tweets.20150430-223406.json')

tweets[ : 10]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="c/CodMn|Ryb@7N-i~LN6">tweets</variable><variable id="B@`]CGD1A2?+*muAJugJ">twitter_samples</variable></variables><block type="variables_set" id="Q=bB?13U{5~]-mP$F$RJ" x="41" y="200"><field name="VAR" id="c/CodMn|Ryb@7N-i~LN6">tweets</field><value name="VALUE"><block type="varDoMethod" id="5L0KrAA7.ekjugpH0_+("><field name="VAR" id="B@`]CGD1A2?+*muAJugJ">twitter_samples</field><field name="MEMBER">strings</field><data>twitter_samples:strings</data><value name="INPUT"><block type="text" id="@8%*,}Voj8$m/8XlDHQz"><field name="TEXT">tweets.20150430-223406.json</field></block></value></block></value></block><block type="lists_getSublist" id="F1K:OnZ`Jb-a{,KN*$N-" x="39" y="320"><mutation at1="false" at2="true"></mutation><field name="WHERE1">FIRST</field><field name="WHERE2">FROM_START</field><value name="LIST"><block type="variables_get" id="(hu+(?ZE;XU3$kDrGoHU"><field name="VAR" id="c/CodMn|Ryb@7N-i~LN6">tweets</field></block></value><value name="AT2"><block type="math_number" id="aPw(:XHX?/z91(}T0feF"><field name="NUM">10</field></block></value></block></xml>

['RT @KirkKus: Indirect cost of the UK being in the EU is estimated to be costing Britain £170 billion per year! #BetterOffOut #UKIP', 'VIDEO: Sturgeon on post-election deals http://t.co/BTJwrpbmOY', 'RT @LabourEoin: The economy was growing 3 times faster on the day David Cameron became Prime Minister than it is today.. #BBCqt http://t.co…', 'RT @GregLauder: the UKIP east lothian candidate looks about 16 and still has an msn addy http://t.co/7eIU0c5Fm1', "RT @thesundaypeople: UKIP's housing spokesman rakes in £800k in housing benefit from migrants.  http://t.co/GVwb9Rcb4w http://t.co/c1AZxcLh…", 'RT @Nigel_Farage: Make sure you tune in to #AskNigelFarage tonight on BBC 1 at 22:50! #UKIP http://t.co/ogHSc2Rsr2', 'RT @joannetallis: Ed Milliband is an embarrassment. Would you want him representing the UK?!  #bbcqt vote @Conservatives', "RT @abstex: The FT is backing the Tories. On an unrelated note, here's a photo of FT leader writer Jonathan Ford (next to Boris) http://t.c…", "RT @NivenJ

As you can see, each tweet is a little document, i.e. a relatively short string of characters.

## Preprocess the data

### Words

Notice that this is a list of tweets, not a single continuous text like `emma`.

Tokenize each tweet into words and store in `tweetTokens1`.

Once you have `tweetTokens1`, show its first two elements.

In [4]:
tweetTokens1 = [(nltk.word_tokenize(i)) for i in tweets]
print(tweetTokens1[ : 2])

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="Q52-g3_[/R0qdo!:lsik">tweetTokens1</variable><variable id="vm#^l?f3f{nfc06H7ta7">i</variable><variable id="c/CodMn|Ryb@7N-i~LN6">tweets</variable><variable id="]HJ?tF]9lV9p9(|h{fmK">nltk</variable></variables><block type="variables_set" id="Zf$0_!.iE^S~):S2$:Y)" x="29" y="200"><field name="VAR" id="Q52-g3_[/R0qdo!:lsik">tweetTokens1</field><value name="VALUE"><block type="lists_create_with" id="UYo/uYP;mBU`1EuTNCkI"><mutation items="1"></mutation><value name="ADD0"><block type="comprehensionForEach" id="q$bdPlP3#D)qGF]yOK]b"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field><value name="LIST"><block type="variables_get" id="R,a!C~+.Y7ZWe!MBjt#)"><field name="VAR" id="c/CodMn|Ryb@7N-i~LN6">tweets</field></block></value><value name="YIELD"><block type="varDoMethod" id="~s.Dw!@c(8npsi6MpcW^"><field name="VAR" id="]HJ?tF]9lV9p9(|h{fmK">nltk</field><field name="MEMBER">word_tokenize</field><data>nltk:word_tokenize</data><value name="INPUT"><block type="variables_get" id="zvm+ZUd[Eb^)P,Mi4ho7"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field></block></value></block></value></block></value></block></value><next><block type="text_print" id="0|gP)^e):dhC1@z?X$8-"><value name="TEXT"><shadow type="text"><field name="TEXT">abc</field></shadow><block type="lists_getSublist" id="])T^~l.j`_k~cW%AbUcU"><mutation at1="false" at2="true"></mutation><field name="WHERE1">FIRST</field><field name="WHERE2">FROM_START</field><value name="LIST"><block type="variables_get" id="mR7/Rgk95Z.N6t(OitdJ"><field name="VAR" id="Q52-g3_[/R0qdo!:lsik">tweetTokens1</field></block></value><value name="AT2"><block type="math_number" id="Mof?1STzpe4u~KdsI-D="><field name="NUM">2</field></block></value></block></value></block></next></block></xml>

[['RT', '@', 'KirkKus', ':', 'Indirect', 'cost', 'of', 'the', 'UK', 'being', 'in', 'the', 'EU', 'is', 'estimated', 'to', 'be', 'costing', 'Britain', '£170', 'billion', 'per', 'year', '!', '#', 'BetterOffOut', '#', 'UKIP'], ['VIDEO', ':', 'Sturgeon', 'on', 'post-election', 'deals', 'http', ':', '//t.co/BTJwrpbmOY']]


----------------------
**QUESTION:**

Which of these tokens violate what we know about Twitter?

**ANSWER: (click here to edit)**

*@ is part of a handle and should not be separated, otherwise we don't know what words are handles.
The same applies to # for hashtags.*

----------------------

We need to use a Twitter-aware tokenizer for these data.
Fortunately, NLTK has one built in.

Create a `TweetTokenizer` using `nltk` and store it in variable `tweetTokenizer`.

In [6]:
tweetTokenizer = nltk.TweetTokenizer()

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</variable><variable id="]HJ?tF]9lV9p9(|h{fmK">nltk</variable></variables><block type="variables_set" id="M6T0am_1LQ~%mS$Po-v^" x="-4" y="179"><field name="VAR" id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</field><value name="VALUE"><block type="varCreateObject" id="A40ZnNV*DXCia`@$Xj`b"><field name="VAR" id="]HJ?tF]9lV9p9(|h{fmK">nltk</field><field name="MEMBER">TweetTokenizer</field><data>nltk:TweetTokenizer</data></block></value></block></xml>

Now tokenize each tweet using `tweetTokenizer` and store in `tweetTokens2`.

Once you have `tweetTokens1`, show its first two elements.

In [7]:
tweetTokens2 = [(tweetTokenizer.tokenize(i)) for i in tweets]
print(tweetTokens2[ : 2])

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</variable><variable id="vm#^l?f3f{nfc06H7ta7">i</variable><variable id="c/CodMn|Ryb@7N-i~LN6">tweets</variable><variable id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</variable></variables><block type="variables_set" id="Zf$0_!.iE^S~):S2$:Y)" x="29" y="200"><field name="VAR" id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</field><value name="VALUE"><block type="lists_create_with" id="UYo/uYP;mBU`1EuTNCkI"><mutation items="1"></mutation><value name="ADD0"><block type="comprehensionForEach" id="q$bdPlP3#D)qGF]yOK]b"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field><value name="LIST"><block type="variables_get" id="R,a!C~+.Y7ZWe!MBjt#)"><field name="VAR" id="c/CodMn|Ryb@7N-i~LN6">tweets</field></block></value><value name="YIELD"><block type="varDoMethod" id="~s.Dw!@c(8npsi6MpcW^"><field name="VAR" id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</field><field name="MEMBER">tokenize</field><data>nltk:word_tokenize</data><value name="INPUT"><block type="variables_get" id="zvm+ZUd[Eb^)P,Mi4ho7"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field></block></value></block></value></block></value></block></value><next><block type="text_print" id="0|gP)^e):dhC1@z?X$8-"><value name="TEXT"><shadow type="text"><field name="TEXT">abc</field></shadow><block type="lists_getSublist" id="])T^~l.j`_k~cW%AbUcU"><mutation at1="false" at2="true"></mutation><field name="WHERE1">FIRST</field><field name="WHERE2">FROM_START</field><value name="LIST"><block type="variables_get" id="mR7/Rgk95Z.N6t(OitdJ"><field name="VAR" id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</field></block></value><value name="AT2"><block type="math_number" id="Mof?1STzpe4u~KdsI-D="><field name="NUM">2</field></block></value></block></value></block></next></block></xml>

[['RT', '@KirkKus', ':', 'Indirect', 'cost', 'of', 'the', 'UK', 'being', 'in', 'the', 'EU', 'is', 'estimated', 'to', 'be', 'costing', 'Britain', '£', '170', 'billion', 'per', 'year', '!', '#BetterOffOut', '#UKIP'], ['VIDEO', ':', 'Sturgeon', 'on', 'post-election', 'deals', 'http://t.co/BTJwrpbmOY']]


----------------------
**QUESTION:**

In addition to fixing handles and hashtags, what other changes did the new tokenizer introduce?

**ANSWER: (click here to edit)**

*The tokenizer is keeping URLs together, which is good. 
Another change is that the currency symbol for pounds sterling is now separated from 170*

----------------------

### Sentences and words

Now process the tweets one more time, this time using the sentence tokenizer first, followed by `tweetTokenizer`, and print the first two elements.

*Hint*: This is a bit of a mind bender, but you can do this with nested comprehensions.
Keep your list comprehension block exactly the same as before, *except* change the `yield` part by:

- Adding another list comprehension (don't forget to put the comprehension in a list like we've been doing)
- In the new comprehension, iterate over `j` and 
    - Use `nltk.sent_tokenize(i)` as input
    - Use `tweetTokenizer.tokenize(j)` for the yield
    
You should be able to figure out the corresponding blocks from what we've covered; if you do it correctly your code will match the above.

*Hint*: If you ever have problems wrapping your head around a nested comprehension, just store the first comprehension in a variable and then use that variable in the second comprehension. Divide and conquer!

In [23]:
tweetTokens2 = [[(tweetTokenizer.tokenize(j)) for j in (nltk.sent_tokenize(i))] for i in tweets]
print(tweetTokens2[ : 2])

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</variable><variable id="vm#^l?f3f{nfc06H7ta7">i</variable><variable id="c/CodMn|Ryb@7N-i~LN6">tweets</variable><variable id="HEsPK%?1|o7!C+UG!1P(">j</variable><variable id="]HJ?tF]9lV9p9(|h{fmK">nltk</variable><variable id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</variable></variables><block type="variables_set" id="Zf$0_!.iE^S~):S2$:Y)" x="29" y="200"><field name="VAR" id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</field><value name="VALUE"><block type="lists_create_with" id="rdRUoL2(PI03TK*V3Y7y"><mutation items="1"></mutation><value name="ADD0"><block type="comprehensionForEach" id="q$bdPlP3#D)qGF]yOK]b"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field><value name="LIST"><block type="variables_get" id="R,a!C~+.Y7ZWe!MBjt#)"><field name="VAR" id="c/CodMn|Ryb@7N-i~LN6">tweets</field></block></value><value name="YIELD"><block type="lists_create_with" id="uQ6):(y|m7w_B{C-?]nk"><mutation items="1"></mutation><value name="ADD0"><block type="comprehensionForEach" id=",Z=/F!k-Z{GH[K7ZXOpP"><field name="VAR" id="HEsPK%?1|o7!C+UG!1P(">j</field><value name="LIST"><block type="varDoMethod" id="~s.Dw!@c(8npsi6MpcW^"><field name="VAR" id="]HJ?tF]9lV9p9(|h{fmK">nltk</field><field name="MEMBER">sent_tokenize</field><data>nltk:sent_tokenize</data><value name="INPUT"><block type="variables_get" id="zvm+ZUd[Eb^)P,Mi4ho7"><field name="VAR" id="vm#^l?f3f{nfc06H7ta7">i</field></block></value></block></value><value name="YIELD"><block type="varDoMethod" id="}AsNv:.WQ;)SyfZ01D3{"><field name="VAR" id="=tk`p3@7pVEk8zIAf7@;">tweetTokenizer</field><field name="MEMBER">tokenize</field><data>nltk:word_tokenize</data><value name="INPUT"><block type="variables_get" id="%q)y5nk]v0v48h6-$(7j"><field name="VAR" id="HEsPK%?1|o7!C+UG!1P(">j</field></block></value></block></value></block></value></block></value></block></value></block></value><next><block type="text_print" id="0|gP)^e):dhC1@z?X$8-"><value name="TEXT"><shadow type="text"><field name="TEXT">abc</field></shadow><block type="lists_getSublist" id="])T^~l.j`_k~cW%AbUcU"><mutation at1="false" at2="true"></mutation><field name="WHERE1">FIRST</field><field name="WHERE2">FROM_START</field><value name="LIST"><block type="variables_get" id="mR7/Rgk95Z.N6t(OitdJ"><field name="VAR" id="uocLxC)3{l9!x]6gHkBp">tweetTokens2</field></block></value><value name="AT2"><block type="math_number" id="Mof?1STzpe4u~KdsI-D="><field name="NUM">2</field></block></value></block></value></block></next></block></xml>

[[['RT', '@KirkKus', ':', 'Indirect', 'cost', 'of', 'the', 'UK', 'being', 'in', 'the', 'EU', 'is', 'estimated', 'to', 'be', 'costing', 'Britain', '£', '170', 'billion', 'per', 'year', '!'], ['#BetterOffOut', '#UKIP']], [['VIDEO', ':', 'Sturgeon', 'on', 'post-election', 'deals', 'http://t.co/BTJwrpbmOY']]]


Notice that we now have a list of tweets, where each tweet is now a list of sentences, and each sentence is a list of word tokens (Twitter-style).

----------------------
**QUESTION:**

Do the sentences in the first two tweets make sense?

**ANSWER: (click here to edit)**

*Somewhat. It is a bit odd that the second 'sentence' of the first tweet is just some hashtags and not a 'real' sentence.
If we wanted to do additional analysis that assumed well-formed sentences, we'd need to make some changes.*

----------------------